File: plateform/frontend/src/types/assets.ts
import type { IncidentClassification, IncidentStatus, SeverityTier } from "./incidents";

export type AssetType = "sensor" | "valve" | "gateway";

export type OperationalStatus = "online" | "degraded" | "offline";

export type ValvePosition = "open" | "closed" | "partial" | "unknown";

export type ControlMode = "manual" | "remote" | "automatic";

export type MaintenanceFilter = "due" | "upcoming" | "scheduled" | "";

export type AssetSortField =
  | "name"
  | "asset_type"
  | "status"
  | "health_score"
  | "last_seen"
  | "next_maintenance";

export type SortOrder = "asc" | "desc";

export interface AssetFilters {
  search: string;
  asset_type: AssetType | "";
  status: OperationalStatus | "";
  zone: string;
  maintenance: MaintenanceFilter;
  sort_by: AssetSortField;
  sort_order: SortOrder;
}

export interface AssetListSummary {
  total_assets: number;
  online: number;
  degraded: number;
  offline: number;
  maintenance_due: number;
}

export interface RelatedAssetIncident {
  id: string;
  incident_number: string;
  title: string;
  classification: IncidentClassification;
  severity: SeverityTier;
  status: IncidentStatus;
  detected_at: string;
}

export interface AssetSummary {
  id: string;
  external_id: string;
  name: string;
  asset_type: AssetType;
  operational_status: OperationalStatus;
  health_score: number | null;
  zone: string;
  zone_id: string;
  zone_name: string;
  location_label: string | null;
  has_cellular_identity: boolean;
  device_msisdn_masked: string | null;
  pipeline_segment: string | null;
  manufacturer: string | null;
  model: string | null;
  battery_pct: number | null;
  current_position: ValvePosition | null;
  last_seen_at: string | null;
  next_maintenance_at: string | null;
  maintenance_state: string;
}

export interface MeasurementRange {
  quantity: string;
  minimum: number;
  maximum: number;
  unit: string;
}

export interface SensorAttributes {
  measurement_types: string[];
  sampling_interval_seconds: number | null;
  battery_pct: number | null;
  signal_strength_dbm: number | null;
  calibration_date: string | null;
  measurement_ranges: MeasurementRange[];
  supported_units: string[];
}

export interface ValveAttributes {
  current_position: ValvePosition | null;
  control_mode: ControlMode | null;
  failsafe_position: string | null;
  actuation_type: string | null;
  last_command_at: string | null;
}

export interface GatewayAttributes {
  provider: string | null;
  connection_type: string | null;
  protocols: string[];
  signal_strength_dbm: number | null;
  packet_delivery_status: string | null;
}

export interface AssetDetail extends AssetSummary {
  serial_number: string | null;
  firmware_version: string | null;
  installed_at: string | null;
  commissioned_at: string | null;
  latitude: number | null;
  longitude: number | null;
  device_location: {
    zone_id: string;
    zone_name: string;
    location_label: string | null;
    latitude: number | null;
    longitude: number | null;
  };
  last_maintenance_at: string | null;
  control_mode: ControlMode | null;
  signal_strength_dbm: number | null;
  sampling_interval_seconds: number | null;
  sensor: SensorAttributes | null;
  valve: ValveAttributes | null;
  gateway: GatewayAttributes | null;
  related_incidents: RelatedAssetIncident[];
}

export interface AssetListResponse {
  items: AssetSummary[];
  total: number;
  summary: AssetListSummary;
}

export interface AssetIncidentListResponse {
  asset_id: string;
  items: RelatedAssetIncident[];
  total: number;
}

export interface AssetHealthPoint {
  timestamp: string;
  health_score: number;
  battery_pct: number | null;
  signal_strength_dbm: number | null;
  availability_pct: number | null;
  position_confirmed: boolean | null;
  packet_delivery_pct: number | null;
  pressure_kpa: number | null;
  flow_lps: number | null;
  packet_loss_pct: number | null;
  temperature_c: number | null;
}

export interface AssetHealthResponse {
  asset_id: string;
  series_kind: string;
  note: string;
  items: AssetHealthPoint[];
  range: string | null;
  freshness: "fresh" | "stale" | "offline" | null;
  last_reading_at: string | null;
  data_mode: string | null;
}

File: plateform/backend/app/schemas/assets.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from app.schemas.incidents import Classification, IncidentStatus, SeverityTier, SortOrder


class AssetType(str, Enum):
    sensor = "sensor"
    valve = "valve"
    gateway = "gateway"


class OperationalStatus(str, Enum):
    online = "online"
    degraded = "degraded"
    offline = "offline"


class ValvePosition(str, Enum):
    open = "open"
    closed = "closed"
    partial = "partial"
    unknown = "unknown"


class ControlMode(str, Enum):
    manual = "manual"
    remote = "remote"
    automatic = "automatic"


class MaintenanceFilter(str, Enum):
    due = "due"
    upcoming = "upcoming"
    scheduled = "scheduled"


class AssetSortField(str, Enum):
    name = "name"
    asset_type = "asset_type"
    status = "status"
    health_score = "health_score"
    last_seen = "last_seen"
    next_maintenance = "next_maintenance"


class MeasurementRange(BaseModel):
    quantity: str
    minimum: float
    maximum: float
    unit: str


class SensorAttributes(BaseModel):
    measurement_types: list[str] = Field(default_factory=list)
    sampling_interval_seconds: int | None = None
    battery_pct: float | None = None
    signal_strength_dbm: int | None = None
    calibration_date: datetime | None = None
    measurement_ranges: list[MeasurementRange] = Field(default_factory=list)
    supported_units: list[str] = Field(default_factory=list)


class ValveAttributes(BaseModel):
    current_position: ValvePosition | None = None
    control_mode: ControlMode | None = None
    failsafe_position: str | None = None
    actuation_type: str | None = None
    last_command_at: datetime | None = None


class GatewayAttributes(BaseModel):
    provider: str | None = None
    connection_type: str | None = None
    protocols: list[str] = Field(default_factory=list)
    signal_strength_dbm: int | None = None
    packet_delivery_status: str | None = None


class DeviceLocation(BaseModel):
    zone_id: str
    zone_name: str
    location_label: str | None = None
    latitude: float | None = None
    longitude: float | None = None


class RelatedIncidentSummary(BaseModel):
    id: str
    incident_number: str
    title: str
    classification: Classification
    severity: SeverityTier
    status: IncidentStatus
    detected_at: datetime


class AssetSummary(BaseModel):
    id: str
    external_id: str
    name: str
    asset_type: AssetType
    operational_status: OperationalStatus
    health_score: float | None = None
    zone: str
    zone_id: str
    zone_name: str
    location_label: str | None = None
    has_cellular_identity: bool = False
    device_msisdn_masked: str | None = None
    pipeline_segment: str | None = None
    manufacturer: str | None = None
    model: str | None = None
    battery_pct: float | None = None
    current_position: ValvePosition | None = None
    last_seen_at: datetime | None = None
    next_maintenance_at: datetime | None = None
    maintenance_state: str


class AssetDetail(AssetSummary):
    serial_number: str | None = None
    firmware_version: str | None = None
    installed_at: datetime | None = None
    commissioned_at: datetime | None = None
    latitude: float | None = None
    longitude: float | None = None
    device_location: DeviceLocation
    last_maintenance_at: datetime | None = None
    control_mode: ControlMode | None = None
    signal_strength_dbm: int | None = None
    sampling_interval_seconds: int | None = None
    sensor: SensorAttributes | None = None
    valve: ValveAttributes | None = None
    gateway: GatewayAttributes | None = None
    related_incidents: list[RelatedIncidentSummary] = Field(default_factory=list)


class AssetListSummary(BaseModel):
    total_assets: int = Field(ge=0)
    online: int = Field(ge=0)
    degraded: int = Field(ge=0)
    offline: int = Field(ge=0)
    maintenance_due: int = Field(ge=0)


class AssetListResponse(BaseModel):
    items: list[AssetSummary]
    total: int = Field(ge=0)
    summary: AssetListSummary


class AssetIncidentListResponse(BaseModel):
    asset_id: str
    items: list[RelatedIncidentSummary]
    total: int = Field(ge=0)


class AssetHealthPoint(BaseModel):
    timestamp: datetime
    health_score: float
    battery_pct: float | None = None
    signal_strength_dbm: int | None = None
    availability_pct: float | None = None
    position_confirmed: bool | None = None
    packet_delivery_pct: float | None = None
    pressure_kpa: float | None = None
    flow_lps: float | None = None
    packet_loss_pct: float | None = None
    temperature_c: float | None = None


class AssetHealthResponse(BaseModel):
    asset_id: str
    series_kind: str = "mock_recent_health"
    note: str
    items: list[AssetHealthPoint]
    range: str | None = None
    freshness: str | None = None
    last_reading_at: datetime | None = None
    data_mode: str | None = None
